### DISCUSSION TITLE
Revisit the minimum supported Python version

### DISCUSSION THREAD

**Discussion by frenck** [home-assistant/architecture#903] [General] [answered]
## Context

As per [ADR-0002](https://github.com/home-assistant/architecture/blob/master/adr/0002-minimum-supported-python-version.md), Home Assistant currently supports the last two minor versions of Python, which are currently (at the time of writing) Python 3.10 and Python 3.11. However, the proposal is to refresh, revisit, or replace ADR-0002, written in May 2019, due to changes in the project's usage and distribution over the years.

> **Note**
> For the purposes of this proposal, it is assumed that Python 3.11 has already been shipped in Home Assistant 2023.6, which is currently in development.

Looking at our current stats, 96% of our users use an installation method that uses one of our pre-built containers, and thus all use the Python 3.11 version we ship bundled in our containers. This means 4% of users may be using either Python 3.10 or Python 3.11, but how this group is divided is unknown.

![Installation methods distribution piechart](https://github.com/home-assistant/architecture/assets/195327/76267d0a-3a44-42bd-a249-c31a2fd5a3d7)

As we maintain support for two versions, the CI tests against both versions, and our dev containers are based on the latest lowest supported version (3.10.0). The reason for the latter is to prevent newer Python features from sneaking in and causing compatibility issues. Although there is a lot of testing and linting in the CI, it does not cover everything.

The introduction of Python 3.11 presents new issues, like, some coroutines need to be wrapped into `asyncio.create_task`. To address this, it would be ideal for running dev containers against Python 3.11 to run into/find and resolve these issues during development. However, maintaining compatibility with Python 3.10, as defined in ADR-0002 is also important.

## Proposal

Considering the dramatic change/shift in our user base from Core-based installs (like Hassbian in the past) to installation methods that rely on our pre-build containers, I want to propose changing our Python version policies:

>  Home Assistant supports a single minor version of Python at a time, with the goal of supporting the latest minor upstream Python version.
>   
> When a new minor Python version is released, the Home Assistant project will provide compatibility and support for that new version as soon as possible.
> 
> Once the new version is supported by the Home Assistant project and rolled out in our container distributions, the previous Python version will be deprecated for a period of two release cycles. Home Assistant will print a warning message during the deprecation period to inform the user.
> 
> Development during the deprecation period will take place on the newer Python version. We rely on automated tests and linting to catch backward incompatible changes.
>
> The minimal supported micro/patch Python version is decided on a minor version basis and codified in the version check code in Home Assistant proper.
    

Some changes in the suggested proposal compared to ADR-0002:

- Only support a single Python version, with a 2-month deprecation/transition period in supporting two version.
- Add clarity on when the deprecation period starts: When shipping our containers with the newer version.
- Add clarity on the intention to upgrade to a new Python version as soon as possible.

## Consequences

- The Python version in our dev container will be quicker matched with 96% of our user base.
- Users of the Core installation type need to upgrade Python at least once every year instead of once every two years. As we already haven’t followed Linux distributions since ADR-0002, those would likely already depend on different Python setups with, for example, `pyenv`.
- Fewer resources are needed for CI. We only need to test with multiple Python versions during the deprecation period.
- We can leverage new Python version features sooner.

## References

- [ADR-0002: **Minimum supported Python version**](https://github.com/home-assistant/architecture/blob/master/adr/0002-minimum-supported-python-version.md)
- Previous discussion: **[Support latest 2 Python versions](https://github.com/home-assistant/architecture/issues/167)**
- Previous discussion: **[Clarify stance on Python micro version support](https://github.com/home-assistant/architecture/issues/278)**

**Comment by thecode** [comment]
I think this will help maintain the project better and support the proposal, but I would suggest to expand the depreciation period to 4 months instead of 2.

**Reply by frenck** [reply]
I've had a call with Erik, Martin & Paulus about this ADR and we have discussed this specific comment as well.

We agreed on the sentiment but would like to stick to the 2 release cycles (months). We will clarify in the ADR, this means 3 cycles (months) of support. During the 3rd month, we still have to support the previous version for backports and patch releases.

**Comment by parautenbach** [comment]
As a core user, an annual Python upgrade wouldn't be too bad. It's more convenient to do it less, but it's typically not a hard exercise (at least it should not be for people that have decided to run HA Core). An OS (Debian) upgrade is a bigger deal, and some Python upgrades to hand-in-hand with OS upgrades (if you want to choose the path of least resistance). 

That said, the Python 3.11 changes are quite significant and not every future version may be like that. Simplifying the build pipelines is likely not a lot of work and I see the general benefits, so it's probably not an unnecessary effort for you, but I wanted to point out this could be a once-off occurrence (the major Python changes).

**Reply by frenck** [reply]
> An OS (Debian) upgrade is a bigger deal, and some Python upgrades to hand-in-hand with OS upgrades (if you want to choose the path of least resistance).

We have not followed or considered Python versions shipped with Linux distributions since ADR-0002 was created in 2019. It was one of the goals of ADR-0002.

There are many options available to run Python, most notably at this point (and my personal recommendation) would be using `pyenv`, as it fully decouples your Python installation (and multiple Python versions) from your OS.

**Reply by parautenbach** [reply]
Sure, I was just making a comparison in support of the change, but I do want to respond to this in particular. I typically make an `altinstall` of Python and build the `pyenv` from there – as recommended, but building Python from sources still require certain OS dependencies.

**Comment by MartinHjelmare** [comment]
I'd like to suggest considering a modification to the ADR so that we still run the previous Python version in development during the deprecation period.

This is what we do today but the ADR will allow us to shorten the time we run more than one Python version. The benefit is that developers won't need to worry about using features of the new Python version that isn't available in the previous version during the deprecation period every year. The downside is that users will start running the new Python version while developers are still running the previous version, and the associated risk of not catching problems in the new Python version as fast.

**Reply by balloob** [reply]
If we run the previous version in dev, it means that for 2 (or 4 if other proposal accepted), the dev environment does not match what 95%+ would run in production.

Our CI would still run the old Python version and will warn when unsupported features are used.

**Reply by MartinHjelmare** [reply]
Yes, I've already noted as much in my initial comment.

**Reply by frenck** [reply]
> it means that for 2 (or 4 if other proposal accepted), the dev environment does not match what 95%+ would run in production.

Actually, that would be +1 (so, 3 or 5 months), as during the month after deprecation ended, we run a cycle that still ports back patch releases.

**Reply by frenck** [reply]
As per the agreement in a call Martin, we will switch to the new version in dev once it matches what we ship to the user.

**Comment by frenck** [comment, marked as answer]
Opened up a PR to formalize the ADR:


https://github.com/home-assistant/architecture/pull/911
